src/Symfony/Component/DomCrawler/Crawler.php

```diff
@@ -968,6 +974,7 @@ public function setDefaultNamespacePrefix($prefix)
     public function registerNamespace($prefix, $namespace)
     {
         $this->namespaces[$prefix] = $namespace;
+        $this->cachedNamespaces[$prefix] = $namespace;
```

Comment:
maybe `cachedNamespaces` should contain only the ones loaded from the namespace discovery rather than also duplicating `$this->namespaces`

Reply:
I knew this'll come up. 😆 
Yeah I was not sure about it. I liked more to have all namespaces in one array, but sure a second check on the namespace array would be still cheap. 
It's just another (small) hurdle developers have to taken into account when the prefixes are used in another place. The namespace duplication is a one time adjustment.
The reason I did it like this at the end was that I don't think registering the namespaces manually is a much used feature. I might be wrong here though.

Separating the cache from the manually registered namespaces is also just necessary because the namespaces are not removed in the `clear` call. Otherwise we could build the cache just in the one object too.

Reply:
Well, explicitly registered namespaces are not removed on `clear` on purpose, because they don't depend on the nodes being in the crawler.

Keeping separate lists would only require checking both in `discoverNamespace` indeed, but that's a `O(1)` check as it checks by key, not by iteration. but it would simplify other places where we would not have to remember to avoid loosing entries from `$this->namespaces`

Reply:
Your argument is I should change it because otherwise we possibly need to remember the namespaces on more places (before that problem may appear we have to empty the cached namespaces on another place, which was one of the reasons I left the private function `clearCachedNamespace` in).

My argument to do it the current way was to not need to remember to do double checks for namespaces in all places that might be added in the future. I didn't even raise the performance card for the additional isset check. It's really really negligible.

For me it's hard to weight which implementation detail is more sensitive. It is also really a detail to me anyway and I'll just make the change if you still prefer it and there is no other opinion.

Reply:
but `cachedNamespaces` is about memoizing the work done by `discoverNamespace`. So there is no reason to access it in more places, as these places must use `discoverNamespace` (they cannot assume that the cache is already populated). There is a single place that reads that array (not counting copying it to sub-crawlers)

Reply:
Ok valid point, there should be no reason to check for namespaces outside of the discoverNamespace function. I'll change it later.